enterprise/ha/src/main/java/org/neo4j/kernel/ha/lock/SlaveLocksClient.java

```diff
@@ -242,9 +247,21 @@ public int getLockSessionId()
     }
 
     @Override
-    public Collection<Locks.ActiveLock> activeLocks()
+    public Stream<? extends ActiveLock> activeLocks()
     {
-        throw new UnsupportedOperationException( "not implemented" );
+        return Stream.concat( // TODO: do we need to make these maps of locks ConcurrentHashMaps?
+                exclusiveLocks.entrySet().stream().flatMap( EXCLUSIVE_ACTIVE_LOCKS ),
+                sharedLocks.entrySet().stream().flatMap( SHARED_ACTIVE_LOCKS ) );
+    }
```

Comment:
Ouch, that's a good point. HashMap is not safe for reads concurrent with writes, because a rehash will reverse the linked list collision buckets, which a concurrent reader than observe as a cycle in the chain, and thus end up in an infinite loop. `ConcurrentHashMap` has slower writes and uses more memory, but I think we probably need it for the lock maps. The maps that segregate by resource type could be turned into arrays and indexed by the resource types ordinal. The probability of seeing an old value in these arrays are probably negligible. It's unfortunate that there's no `AtomicReferenceArrayFieldUpdater`.